Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/TapToPlace.cs

```diff
@@ -0,0 +1,344 @@
+        private bool keepOrientationVertical = false;
+
+        /// <summary>
+        /// If true, the GameObjectToPlace will remain upright and in line with Vector3.up
+        /// </summary>
+        public bool KeepOrientationVertical
+        {
+            get => keepOrientationVertical;
+            set => keepOrientationVertical = value;
+        }
+
+        [SerializeField]
+        [Tooltip("If true and in the Unity Editor, the normal of the raycast hit will be drawn in yellow.")]
+        private bool debugEnabled = true;
+
+        /// <summary>
+        /// If true and in the Unity Editor, the normal of the raycast hit will be drawn in yellow.
+        /// </summary>
+        public bool DebugEnabled
+        {
+            get => debugEnabled;
+            set => debugEnabled = value;
+        }
+
+        [SerializeField]
+        [Tooltip("If false, the game object to place will not change its rotation according to the surface hit.  The object will" +
+            " remain facing the camera while it is in the placing state.  If true, the object will rotate according to the surface normal" +
+            " if there is a hit.")]
+        private bool rotateAccordingToSurface = false;
+
+        /// <summary>
+        /// If false, the game object to place will not change its rotation according to the surface hit.  The object will 
+        /// remain facing the camera while it is in the placing state.  If true, the object will rotate according to the surface normal
+        /// if there is a hit."
+        /// </summary>
+        public bool RotateAccordingToSurface
+        {
+            get => rotateAccordingToSurface;
+            set => rotateAccordingToSurface = value;
+        }
+
+        [Tooltip("This event is triggered once at the start of the placing state.")]
+        /// <summary>
+        /// This event is triggered once at the start of the placing state.
+        /// </summary>
+        public UnityEvent OnPlacingStarted = new UnityEvent();
+
+        [Tooltip("This event is triggered once when the placing state has ended.")]
+        /// <summary>
+        /// This event is triggered once when the placing state has ended.
+        /// </summary>
+        public UnityEvent OnPlacingStopped = new UnityEvent();
+
+        /// <summary>
+        /// Get the game object layer before the object is placed.  When an object is in the placing state, the layer
+        /// is changed to IgnoreRaycast and then back to the original layer.
+        /// </summary>
+        public int GameObjectLayer { get; protected set; }
+
+        protected internal bool IsColliderPresent => gameObject != null ? gameObject.GetComponent<Collider>() != null : false;
+
+        private const int IgnoreRaycastLayer = 2;
+
+        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint)
+        protected RayStep CurrentRay;
```

Comment:
Nit: If you have a set of data that is being mutated once per frame (or is otherwise used to avoid passing extra state between functions like we have with DidHitSurface) it can help to add some comments around them (i.e. the below properties are used in the same frame to compute hits and such

Reply:
Grouped the Ray properties and added comment:
```
        // The current ray is based on the TrackedTargetType (Controller Ray, Head, Hand Joint).
        // The following properties are updated each frame while the game object is selected to determine
        // object placement if there is a hit on a surface.
        protected RayStep CurrentRay;
        protected bool DidHitSurface;
        protected RaycastHit CurrentHit;
```